File: services/binanceService.js
const axios = require('axios');
require('dotenv').config(); // Make sure env vars are loaded
const logger = require('../config/logger');

const BNS_UUID = process.env.BNS_UUID;
const API_ENDPOINT = 'https://www.binance.com/bapi/futures/v1/friendly/future/copy-trade/lead-portfolio/trade-history';

async function fetchTradeHistory(portfolioId, page = 1, pageSize = 100) {
  if (!BNS_UUID) {
    logger.error('BNS_UUID is not configured.');
    throw new Error('BNS_UUID is not configured.');
  }
  logger.info(`Fetching trade history for portfolioId: ${portfolioId}, pageNumber: ${page}, pageSize: ${pageSize}`);
  try {
    const payload = {
      portfolioId: String(portfolioId),
      pageNumber: Number(page), // Changed from page to pageNumber
      pageSize: Number(pageSize),
      tradeType: "ALL"
    };
    // The subtask for time-windowed fetching will add startTime and endTime to this payload.
    // For now, this is the correct structure based on the current subtask.

    const response = await axios.post(API_ENDPOINT, payload,
      { headers: { 'bns-uuid': BNS_UUID, 'Content-Type': 'application/json' } }
    );

    if (response.data && response.data.code === "000000") {
      logger.info(`Successfully fetched ${response.data.data.list.length} trades for portfolioId: ${portfolioId}, pageNumber: ${page}. Total items: ${response.data.data.total}`);
      return response.data.data; // Contains list and total
    } else {
      logger.error(`API error for portfolioId ${portfolioId}, pageNumber: ${page}: ${response.data.message || 'Unknown API error'}`, response.data);
      throw new Error(`Binance API error: ${response.data.message || 'Unknown error'}`);
    }
  } catch (error) {
    logger.error(`Failed to fetch trade history for portfolioId ${portfolioId}, pageNumber: ${page}: ${error.message}`, { stack: error.stack });
    if (error.response) {
      logger.error('Error response data:', error.response.data);
      logger.error('Error response status:', error.response.status);
    }
    throw error; // Re-throw to be handled by the caller
  }
}

async function fetchAllTrades(portfolioId) {
  let allTrades = [];
  let currentPage = 1;
  const pageSize = 100; // Max pageSize by API seems to be 100
  let totalPages = 1; 
  let retries = 0;
  const maxRetries = 3;

  logger.info(`Starting to fetch all trades for portfolioId: ${portfolioId}`);

  do {
    try {
      const data = await fetchTradeHistory(portfolioId, currentPage, pageSize);
      if (data && data.list) {
        allTrades = allTrades.concat(data.list);
        // Ensure total is a number and greater than 0 before calculating totalPages
        const totalTrades = Number(data.total);
        totalPages = totalTrades > 0 ? Math.ceil(totalTrades / pageSize) : 0;

        logger.info(`Fetched page ${currentPage}/${totalPages === 0 && totalTrades === 0 ? 1 : totalPages} for portfolioId ${portfolioId}. Total trades so far: ${allTrades.length}`);
        
        if (totalPages === 0) { // No trades to fetch or already fetched all
            break;
        }
        currentPage++;
        retries = 0; // Reset retries on success
      } else {
        // This case should ideally be caught by error handling in fetchTradeHistory
        logger.warn(`No data received for portfolioId ${portfolioId}, page ${currentPage}, but no explicit error thrown.`);
        // Consider a retry or break strategy here
        break; 
      }
    } catch (error) {
      logger.error(`Error fetching page ${currentPage} for portfolioId ${portfolioId}: ${error.message}. Attempt ${retries + 1}/${maxRetries}`);
      retries++;
      if (retries >= maxRetries) {
        logger.error(`Max retries reached for portfolioId ${portfolioId}, page ${currentPage}. Aborting for this portfolio.`);
        // Decide: throw error to stop all processing, or just return what's been collected.
        // For now, let's throw to indicate a significant issue.
        throw new Error(`Failed to fetch all trades for ${portfolioId} after ${maxRetries} retries on page ${currentPage}.`);
      }
      await new Promise(resolve => setTimeout(resolve, 2000 * retries)); // Wait longer after each retry
      continue; // Retry the current page
    }

    // Delay between successful page fetches
    if (currentPage <= totalPages) {
      await new Promise(resolve => setTimeout(resolve, 1000)); // 1-second delay
    }
  } while (currentPage <= totalPages && totalPages > 0); // Continue if there are more pages to fetch

  logger.info(`Successfully fetched all ${allTrades.length} trades for portfolioId: ${portfolioId} over ${totalPages === 0 && allTrades.length === 0 ? 1 : totalPages} pages.`);
  return allTrades;
}

module.exports = { fetchAllTrades, fetchTradeHistory }; // Export both if fetchTradeHistory might be useful externally or for testing

File: services/binance_service.py
import requests
import logging
import time # Potentially for rate limiting

BINANCE_API_URL = "https://www.binance.com/bapi/futures/v1/friendly/future/copy-trade/lead-portfolio/trade-history"
DEFAULT_PAGE_SIZE = 100

def fetch_trade_history(portfolio_id: str, bns_uuid: str, page_size: int = DEFAULT_PAGE_SIZE) -> list | None:
    logger = logging.getLogger(__name__)
    all_trades = []
    page_no = 1
    
    # Log partial bns_uuid for privacy/security
    bns_uuid_display = bns_uuid[:10] + "..." if bns_uuid and len(bns_uuid) > 10 else "INVALID_BNS_UUID"
    logger.info(f"Starting to fetch trade history for portfolioId: {portfolio_id} with bns_uuid: {bns_uuid_display}")

    while True:
        payload = {
            "portfolioId": portfolio_id,
            "pageNo": page_no,
            "pageSize": page_size
        }
        headers = {
            "bns-uuid": bns_uuid,
            "Content-Type": "application/json",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
        }

        try:
            logger.debug(f"Fetching page {page_no} for portfolioId {portfolio_id} with payload: {payload}")
            response = requests.post(BINANCE_API_URL, headers=headers, json=payload, timeout=10) # 10 seconds timeout

            # Check for HTTP errors first (4xx or 5xx)
            if response.status_code != 200:
                logger.error(f"HTTP error {response.status_code} while fetching trades for portfolioId {portfolio_id} (page {page_no}): {response.text}")
                return None # Indicate failure

            data = response.json()

            if data.get("code") != "000000":
                logger.error(f"Binance API error for portfolioId {portfolio_id}: Code {data.get('code')}, Message: {data.get('message')}")
                return None # Indicate failure due to API error code

            current_page_trades = data.get("data", {}).get("list", [])
            if not current_page_trades:
                logger.info(f"No more trades found for portfolioId {portfolio_id} on page {page_no}. This might be the end of data or an empty page.")
                break # No more data on this page, assume end of records for this portfolio

            all_trades.extend(current_page_trades)
            total_records_api = data.get("data", {}).get("total", 0)
            
            logger.debug(f"Fetched {len(current_page_trades)} trades on page {page_no} for portfolioId {portfolio_id}. Total fetched so far: {len(all_trades)} out of {total_records_api} (API reported).")

            # Check if all records have been fetched
            if len(all_trades) >= total_records_api or len(current_page_trades) < page_size:
                logger.info(f"Completed fetching all trades for portfolioId {portfolio_id}. Total fetched: {len(all_trades)} (API reported {total_records_api}).")
                break 
            
            page_no += 1
            # time.sleep(0.5) # Optional: brief pause to be polite to the API; uncomment if rate limiting is an issue

        except requests.exceptions.Timeout:
            logger.error(f"Timeout occurred while fetching trades for portfolioId {portfolio_id} (page {page_no}).")
            return None # Indicate failure
        except requests.exceptions.RequestException as e: # Catches other requests-related errors (network, etc.)
            logger.error(f"RequestException fetching trades for portfolioId {portfolio_id} (page {page_no}): {e}")
            return None # Indicate failure
        except Exception as e: # Catch any other unexpected error during processing (e.g., JSON parsing if not valid JSON)
            logger.error(f"An unexpected error occurred during trade fetching for portfolioId {portfolio_id} (page {page_no}): {e}", exc_info=True)
            return None


    logger.info(f"Successfully fetched a total of {len(all_trades)} trades for portfolio ID {portfolio_id}.")
    return all_trades

# Example usage (optional, for testing by worker, ensure .env is set up)
# if __name__ == '__main__':
#     # This requires a valid BNS_UUID and PORTFOLIO_ID in .env or passed directly
#     # Ensure utils.logger and config are available.
#     from dotenv import load_dotenv
#     import os
#     # Assuming this script is in services/ and utils/ is a sibling directory
#     import sys
#     sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
#     from utils.logger import setup_logging # Import your setup_logging function
    
#     load_dotenv() # Load .env file from project root

#     # Setup logging for the test
#     setup_logging(logging.DEBUG) 
#     test_logger = logging.getLogger(__name__) # Get logger for this module

#     BNS_UUID_TEST = os.getenv("BNS_UUID")
#     PORTFOLIO_IDS_STR = os.getenv("PORTFOLIO_IDS")
    
#     if not BNS_UUID_TEST:
#         test_logger.error("BNS_UUID not found in environment variables.")
#     if not PORTFOLIO_IDS_STR:
#         test_logger.error("PORTFOLIO_IDS not found in environment variables.")

#     if BNS_UUID_TEST and PORTFOLIO_IDS_STR:
#         PORTFOLIO_ID_TEST = PORTFOLIO_IDS_STR.split(',')[0].strip() # Get the first portfolio ID
#         if PORTFOLIO_ID_TEST:
#             test_logger.info(f"Attempting to fetch trades for PORTFOLIO_ID: {PORTFOLIO_ID_TEST} with BNS_UUID starting with: {BNS_UUID_TEST[:10]}")
#             trades_result = fetch_trade_history(PORTFOLIO_ID_TEST, BNS_UUID_TEST, page_size=10) # Small page size for testing
#             if trades_result is not None:
#                 test_logger.info(f"Fetched {len(trades_result)} trades successfully.")
#                 # You can uncomment the following to print some trade details:
#                 # for i, trade in enumerate(trades_result[:3]): # Print first 3 trades
#                 #     test_logger.info(f"Trade {i+1}: {trade.get('symbol')}, Qty: {trade.get('qty')}, Price: {trade.get('price')}, Time: {trade.get('time')}")
#             else:
#                 test_logger.error(f"Failed to fetch trades for {PORTFOLIO_ID_TEST}.")
#         else:
#             test_logger.error("No portfolio ID found after splitting PORTFOLIO_IDS.")
#     else:
#         test_logger.warning("BNS_UUID or PORTFOLIO_IDS not configured in .env for testing.")
